File: lib/games/tag/tag_transport.dart
import 'dart:async';

import 'package:web_socket_channel/web_socket_channel.dart';

import '../../social/host_server.dart';
import 'tag_protocol.dart';

/// Bidirectional channel for tag messages. The host wraps its in-app
/// `HostServer` (fan-out to all connected app peers). Each peer wraps a
/// single outbound WebSocket to the host.
abstract class TagTransport {
  /// Outbound: send to everyone (host) or to the host (peer).
  void send(TagMessage msg);

  /// Inbound: TagMessages decoded from the wire.
  Stream<TagMessage> get inbound;

  /// Optional: who just connected. Host emits one event per app-peer
  /// WebSocket join. Peer transport never emits.
  Stream<String> get onPeerConnected;

  /// Optional: who just disconnected. Mirror of onPeerConnected.
  Stream<String> get onPeerDisconnected;

  Future<void> dispose();
}

/// Host-side: wraps a running HostServer. All inbound guest messages are
/// parsed as TagMessages; `send` broadcasts to every connected guest.
/// Owns the server — `dispose()` stops it.
class HostTagTransport implements TagTransport {
  HostTagTransport(this._server) {
    _msgSub = _server.onGuestMessage.listen(_onGuestMessage);
    _joinSub = _server.onGuestJoin.listen((g) => _onConnect.add(g.value));
    _leaveSub = _server.onGuestLeave.listen((g) {
      final pid = _guestToPeer.remove(g);
      if (pid != null) _onDisconnect.add(pid);
    });
  }

  final HostServer _server;
  final _inbound = StreamController<TagMessage>.broadcast();
  final _onConnect = StreamController<String>.broadcast();
  final _onDisconnect = StreamController<String>.broadcast();
  final Map<GuestId, String> _guestToPeer = {};

  StreamSubscription? _msgSub;
  StreamSubscription? _joinSub;
  StreamSubscription? _leaveSub;

  void _onGuestMessage(GuestMessage m) {
    try {
      final msg = TagMessage.decode(m.payload);
      if (msg is HelloMessage) {
        _guestToPeer[m.from] = msg.peerId;
      }
      _inbound.add(msg);
      // Echo non-start traffic back so other peers see it. Host's
      // TagSession already applied any state change locally.
      if (msg is! HelloMessage) {
        _server.broadcast(m.payload);
      }
    } catch (_) {
      // ignore malformed payloads
    }
  }

  @override
  void send(TagMessage msg) => _server.broadcast(msg.encode());

  @override
  Stream<TagMessage> get inbound => _inbound.stream;
  @override
  Stream<String> get onPeerConnected => _onConnect.stream;
  @override
  Stream<String> get onPeerDisconnected => _onDisconnect.stream;

  @override
  Future<void> dispose() async {
    await _msgSub?.cancel();
    await _joinSub?.cancel();
    await _leaveSub?.cancel();
    await _server.stop();
    await _inbound.close();
    await _onConnect.close();
    await _onDisconnect.close();
  }
}

/// Peer-side: connects one WebSocket to the host. Everything `send`-ed
/// goes to the host, which fans out to all peers (including us). Inbound
/// stream surfaces the fanned-out copies.
class PeerTagTransport implements TagTransport {
  PeerTagTransport(this._channel) {
    _sub = _channel.stream.listen(
      (data) {
        if (data is String) {
          try {
            _inbound.add(TagMessage.decode(data));
          } catch (_) {
            // skip
          }
        }
      },
      onError: (_) {},
      onDone: () {},
      cancelOnError: false,
    );
  }

  factory PeerTagTransport.connect(Uri serverUri) {
    final wsUri = serverUri.replace(
      scheme: serverUri.scheme == 'https' ? 'wss' : 'ws',
      path: '/ws',
    );
    return PeerTagTransport(WebSocketChannel.connect(wsUri));
  }

  final WebSocketChannel _channel;
  late final StreamSubscription _sub;
  final _inbound = StreamController<TagMessage>.broadcast();

  @override
  void send(TagMessage msg) => _channel.sink.add(msg.encode());

  @override
  Stream<TagMessage> get inbound => _inbound.stream;

  @override
  Stream<String> get onPeerConnected => const Stream<String>.empty();
  @override
  Stream<String> get onPeerDisconnected => const Stream<String>.empty();

  @override
  Future<void> dispose() async {
    await _sub.cancel();
    await _channel.sink.close();
    await _inbound.close();
  }
}

/// Local-only transport. Used when there's no network (single-device dev
/// mode). Messages echo back into [inbound] so the engine still sees
/// what it sent.
class LoopbackTagTransport implements TagTransport {
  final _inbound = StreamController<TagMessage>.broadcast();
  @override
  void send(TagMessage msg) => _inbound.add(msg);
  @override
  Stream<TagMessage> get inbound => _inbound.stream;
  @override
  Stream<String> get onPeerConnected => const Stream<String>.empty();
  @override
  Stream<String> get onPeerDisconnected => const Stream<String>.empty();
  @override
  Future<void> dispose() async => _inbound.close();
}

File: android/app/src/main/java/com/example/jamboree/games/tag/TagTransport.kt
package com.example.jamboree.games.tag

import com.example.jamboree.join.GuestLink
import com.example.jamboree.social.GuestId
import com.example.jamboree.social.HostServer
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject

/**
 * Bidirectional channel for tag messages. The host wraps its in-app
 * [HostServer]; each peer wraps a single outbound WebSocket to the host.
 */
interface TagTransport {
    var onInbound: ((TagMessage) -> Unit)?
    var onPeerConnected: ((String) -> Unit)?
    var onPeerDisconnected: ((String) -> Unit)?
    fun send(msg: TagMessage)
    fun dispose()
}

/** Host-side: wraps a running [HostServer]. */
class HostTagTransport(private val server: HostServer) : TagTransport {
    private val guestToPeer = HashMap<GuestId, String>()
    /** peerId → chosen display name, learned from the join handshake and the
     *  peer's `hello`. The lobby uses this so peers show their real name. */
    private val peerDisplayNames = HashMap<String, String>()

    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null

    fun displayName(peerId: String): String = peerDisplayNames[peerId] ?: peerId

    /** Whether a peer-originated message is allowed from [sender] (the peerId
     *  bound to that connection at `hello`/join). `Tag`/`Unfreeze`/
     *  `TutorialVote` are only honored when the actor id matches the sender;
     *  `Start`/`End`/`TutorialState` are host-authoritative and never accepted
     *  from a peer. `Hello` is handled before this check. */
    private fun isAuthentic(msg: TagMessage, sender: String?): Boolean = when (msg) {
        is TagMessage.Tag -> sender != null && msg.taggerId == sender
        is TagMessage.Unfreeze -> sender != null && msg.unfreezerId == sender
        is TagMessage.TutorialVote -> sender != null && msg.voterId == sender
        is TagMessage.TutorialCall -> true
        is TagMessage.Start, is TagMessage.End,
        is TagMessage.TutorialState, is TagMessage.Hello -> false
    }

    init {
        server.onJoin = { g -> onPeerConnected?.invoke(g.value) }
        server.onLeave = { g ->
            guestToPeer.remove(g)?.let {
                peerDisplayNames.remove(it)
                onPeerDisconnected?.invoke(it)
            }
        }
        server.onMessage = { g, raw ->
            val msg = runCatching { TagMessage.decode(raw) }.getOrNull()
            if (msg != null) {
                if (msg is TagMessage.Hello) {
                    guestToPeer[g] = msg.peerId
                    peerDisplayNames[msg.peerId] = msg.displayName
                    onInbound?.invoke(msg)  // hello is not relayed to other peers
                } else if (isAuthentic(msg, guestToPeer[g])) {
                    onInbound?.invoke(msg)
                    // Echo to other peers. The host's own TagSession already
                    // applied any state change locally.
                    server.broadcast(raw)
                }
                // Else: a spoofed actor or a peer-forged round-control/state
                // message. The host is the authoritative relay, so drop it
                // here — it is neither applied on the host nor echoed, so no
                // peer can tag/unfreeze on behalf of another player or end the
                // round with an arbitrary winner.
            } else {
                // Not a TagMessage — the browser-tier "Join a game"
                // handshake. App peers join Tag by code: complete the
                // handshake so the generic flow can mount the native Tag
                // peer screen, then the peer speaks TagMessages over this
                // same socket.
                val j = runCatching { JSONObject(raw) }.getOrNull()
                if (j?.optString("type") == "join") {
                    val name = j.optString("name").trim()
                    val peerId = g.value
                    guestToPeer[g] = peerId
                    peerDisplayNames[peerId] = name.ifEmpty { peerId }
                    server.send(g, JSONObject()
                        .put("type", "welcome").put("game", "tag")
                        .put("yourId", peerId).put("yourName", name).toString())
                } else {
                    server.disconnect(g,
                        """{"type":"error","message":"This host is running Tag — open it in the app to join."}""")
                }
            }
        }
    }

    override fun send(msg: TagMessage) = server.broadcast(msg.encode())
    override fun dispose() = server.stopServer()
}

/** Peer-side: connects one WebSocket to the host. */
class PeerTagTransport private constructor(serverUrl: String) : TagTransport {
    private val client = OkHttpClient()
    private var ws: WebSocket? = null

    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null

    init {
        val req = Request.Builder().url(serverUrl).build()
        ws = client.newWebSocket(req, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                runCatching { onInbound?.invoke(TagMessage.decode(text)) }
            }
        })
    }

    companion object {
        /** Convert "https://host:port/" to "wss://host:port/ws" and connect. */
        fun connect(serverUrl: String): PeerTagTransport {
            val wsUrl = serverUrl
                .replaceFirst("https://", "wss://")
                .replaceFirst("http://", "ws://")
                .trimEnd('/') + "/ws"
            return PeerTagTransport(wsUrl)
        }
    }

    override fun send(msg: TagMessage) { ws?.send(msg.encode()) }
    override fun dispose() {
        ws?.close(1000, null); ws = null
        client.dispatcher.executorService.shutdown()
    }
}

/**
 * App-peer transport that rides the browser-tier [GuestLink] socket the
 * generic "Join a game" flow already opened, rather than dialing a second
 * raw WebSocket. Tag messages travel as the same JSON every other game uses.
 */
class GuestLinkTagTransport(private val client: GuestLink) : TagTransport {
    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null

    /** Subscribe to the socket. Call this *after* the owning [TagSession]
     *  has set [onInbound]: assigning [GuestLink.onMessage] synchronously
     *  flushes frames buffered since `welcome`, which would otherwise
     *  arrive before the session is wired. */
    fun start() {
        client.onMessage = { j ->
            runCatching { TagMessage.decode(j) }.getOrNull()?.let { onInbound?.invoke(it) }
        }
    }

    override fun send(msg: TagMessage) { client.send(msg.toJson()) }
    override fun dispose() { client.onMessage = null }
}

/** Local-only transport for single-device dev mode. */
class LoopbackTagTransport : TagTransport {
    override var onInbound: ((TagMessage) -> Unit)? = null
    override var onPeerConnected: ((String) -> Unit)? = null
    override var onPeerDisconnected: ((String) -> Unit)? = null
    override fun send(msg: TagMessage) { onInbound?.invoke(msg) }
    override fun dispose() {}
}
